src/typescript/frontend/src/lib/store/name-store.ts

```diff
@@ -0,0 +1,173 @@
+
+  const serializedMap = JSON.stringify(entries);
+
+  localStorage.setItem(LOCALSTORAGE_ANS_KEY, serializedMap);
+}
+
+/** Loads the ANSMap from localStorage. */
+export function loadANSMap(): ANSMap {
+  const map = localStorage.getItem(LOCALSTORAGE_ANS_KEY);
+  if (map) {
+    return new Map(JSON.parse(map));
+  } else {
+    return new Map();
+  }
+}
+
+/**
+ * Get the name of an address.
+ *
+ * Returns null if the address has no registered name.
+ *
+ * Do not expose this function to avoid using it in a for loop.
+ */
+async function fetchANSName(address: string): Promise<string | null> {
+  const primaryNameUrl = getRequestURL(address, true);
+
+  try {
+    const { name: primaryName } = await fetch(primaryNameUrl).then((res) => res.json());
+
+    if (primaryName) {
+      return primaryName;
+    } else {
+      const nameUrl = getRequestURL(address, false);
+
+      if (!nameUrl) {
+        return null;
+      }
+
+      const { name } = await fetch(nameUrl).then((res) => res.json());
+      if (name) {
+        return name;
+      }
+      return null;
+    }
+  } catch (_) {
+    return null;
+  }
+}
+
+export const initializeEventStore = (): NameState => {
+  if (typeof localStorage !== "undefined") {
+    return { names: loadANSMap() };
+  } else {
+    return { names: new Map() };
+  }
+};
+
+export const defaultState: NameState = initializeEventStore();
+
+type Set = (
+  nextStateOrUpdater: NameStore | Partial<NameStore> | ((state: WritableDraft<NameStore>) => void),
+  shouldReplace?: boolean | undefined
+) => void;
+type Get = () => NameStore;
+
+const getNameMap = (set: Set, get: Get, addresses: string[]): ANSMap => {
+  const map = get().names;
+  const uncachedNames: string[] = [];
+  for (const address of addresses) {
+    if (!map.has(address)) {
+      uncachedNames.push(address);
+    }
+  }
+  if (uncachedNames.length > 0) {
+    for (const address of uncachedNames) {
+      if (!ansPromiseMap.has(address)) {
+        const x = fetchANSName(address)
+          .then((res) => ({
+            name: res,
+            isNull: res === null,
+            expiry: new Date().getTime() + ANS_CACHE_TIME,
+          }))
+          .then((res) => {
+            set((state) => {
+              state.names.set(address, res);
+              ansPromiseMap.delete(address);
+              if (ansPromiseMap.size === 0) {
+                saveANSMap(state.names);
+              }
+            });
+          });
+        ansPromiseMap.set(address, x);
+      }
+    }
+  }
+  return get().names;
+};
+
+export const createNameStore = (initialState: NameState = defaultState) => {
+  return createStore<NameStore>()(
+    immer((set, get) => ({
+      ...initialState,
+      getMapWithNames: (addresses) => {
+        return getNameMap(set, get, addresses);
+      },
+      getResolverWithNames: (addresses) => {
+        const map = getNameMap(set, get, addresses);
+        return (address) => map.get(address)?.name ?? address;
+      },
```

Comment:
I think you won't need these anymore if you use the hook instead, btw, all you need really is just the map, which will notify the hook of changes